File: web/src/api/riskAssessment.ts
import request, { rawRequest } from './request'

export interface RiskHit {
  repo?: string
  format?: string
  version?: string
  dependent_name?: string
  constraint?: string
}

export interface RiskAssessmentItem {
  id: number
  assessment_id: number
  name: string
  version: string
  format?: string
  severity?: string
  reason?: string
  cve?: string
  matched: boolean
  artifact_hit?: RiskHit[]
  dependency_hit?: RiskHit[]
  suggestion?: string
  disposition?: '' | 'removed' | 'blocked' | 'ignored'
  disposition_note?: string
  disposed_by?: number
  disposed_at?: string | null
  created_at: string
}

export interface RiskAssessment {
  id: number
  file_name: string
  total_items: number
  matched_items: number
  artifact_hits: number
  dependency_hits: number
  created_by?: number
  created_at: string
  items?: RiskAssessmentItem[]
}

export interface RiskComponent {
  name: string
  version: string
  format?: string
  severity?: string
  reason?: string
  cve?: string
}

export const riskAssessmentApi = {
  upload(file: File) {
    const formData = new FormData()
    formData.append('file', file)
    return request.post<RiskAssessment>('/security/risk-assessments/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120000,
    })
  },

  aiParse(file: File) {
    const formData = new FormData()
    formData.append('file', file)
    return request.post<{ file_name: string; components: RiskComponent[]; count: number }>(
      '/security/risk-assessments/ai-parse',
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 120000,
      },
    )
  },

  analyze(fileName: string, components: RiskComponent[]) {
    return request.post<RiskAssessment>('/security/risk-assessments/analyze', { file_name: fileName, components })
  },

  aiReport(id: number) {
    return request.post<{ report: string }>(`/security/risk-assessments/${id}/ai-report`, {}, { timeout: 120000 })
  },

  createBlockRules(id: number, itemIds: number[]) {
    return request.post<{ count: number }>(`/security/risk-assessments/${id}/block-rules`, { item_ids: itemIds })
  },

  dispose(itemId: number, action: 'remove' | 'block' | 'ignore' | 'reset', note?: string) {
    return request.post<RiskAssessmentItem>('/security/risk-assessments/dispose', {
      item_id: itemId,
      action,
      note: note || '',
    })
  },

  disposeStats(id: number) {
    return request.get<{ total: number; disposed: number; removed: number; blocked: number; ignored: number }>(
      `/security/risk-assessments/${id}/dispose-stats`,
    )
  },

  list(params?: { page?: number; page_size?: number }) {
    return request.get<{ items: RiskAssessment[]; pagination: { total: number; page: number; page_size: number } }>(
      '/security/risk-assessments',
      { params },
    )
  },

  get(id: number) {
    return request.get<RiskAssessment>(`/security/risk-assessments/${id}`)
  },

  downloadTemplate() {
    return rawRequest.get('/security/risk-assessments/template', { responseType: 'blob' })
  },

  exportFile(id: number) {
    return rawRequest.get(`/security/risk-assessments/${id}/export`, { responseType: 'blob' })
  },

  remove(id: number) {
    return request.delete(`/security/risk-assessments/${id}`)
  },
}

File: internal/model/risk_assessment.go
package model

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"
)

// JSONArray 存储 JSON 数组列（如命中明细列表）。
type JSONArray []map[string]interface{}

func (a JSONArray) Value() (driver.Value, error) {
	if a == nil {
		return nil, nil
	}
	return json.Marshal(a)
}

func (a *JSONArray) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}
	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return fmt.Errorf("unsupported JSONArray scan type: %T", value)
	}
	if len(bytes) == 0 {
		*a = nil
		return nil
	}
	return json.Unmarshal(bytes, a)
}

// RiskAssessment 风险组件研判记录（头表）。
// 对应一次 Excel 上传的研判，Items 为明细。
type RiskAssessment struct {
	ID             uint                 `gorm:"primaryKey" json:"id"`
	FileName       string               `gorm:"size:255" json:"file_name"`
	TotalItems     int                  `gorm:"not null;default:0" json:"total_items"`
	MatchedItems   int                  `gorm:"not null;default:0" json:"matched_items"`
	ArtifactHits   int                  `gorm:"not null;default:0" json:"artifact_hits"`
	DependencyHits int                  `gorm:"not null;default:0" json:"dependency_hits"`
	CreatedBy      uint                 `gorm:"index" json:"created_by,omitempty"`
	CreatedAt      time.Time            `gorm:"autoCreateTime" json:"created_at"`
	Items          []RiskAssessmentItem `gorm:"foreignKey:AssessmentID" json:"items,omitempty"`
}

func (RiskAssessment) TableName() string {
	return "risk_assessments"
}

// 处置状态
type DispositionStatus string

const (
	DispositionNone     DispositionStatus = ""            // 未处置
	DispositionRemoved  DispositionStatus = "removed"     // 已移除制品
	DispositionBlocked  DispositionStatus = "blocked"     // 仅生成阻断规则
	DispositionIgnored  DispositionStatus = "ignored"     // 已忽略
)

// RiskAssessmentItem 风险组件研判明细。
// ArtifactHit 形如 [{"repo":"central","format":"maven","version":"1.2.3"}]
// DependencyHit 形如 [{"dependent_name":"my-app","version":"1.0.0","format":"npm","repo":"central","constraint":">=1.0.0"}]
type RiskAssessmentItem struct {
	ID               uint      `gorm:"primaryKey" json:"id"`
	AssessmentID     uint      `gorm:"not null;index" json:"assessment_id"`
	Name             string    `gorm:"size:512;index" json:"name"`
	Version          string    `gorm:"size:255" json:"version"`
	Format           string    `gorm:"size:32" json:"format,omitempty"`
	Severity         string    `gorm:"size:32" json:"severity,omitempty"`
	Reason           string    `gorm:"size:512" json:"reason,omitempty"`
	CVE              string    `gorm:"size:64" json:"cve,omitempty"`
	Matched          bool      `gorm:"not null;default:false;index" json:"matched"`
	ArtifactHit      JSONArray `gorm:"type:json" json:"artifact_hit,omitempty"`
	DependencyHit    JSONArray `gorm:"type:json" json:"dependency_hit,omitempty"`
	Disposition      DispositionStatus `gorm:"size:20;not null;default:'';index" json:"disposition"`
	DispositionNote  string    `gorm:"size:512" json:"disposition_note,omitempty"`
	DisposedBy       uint      `json:"disposed_by,omitempty"`
	DisposedAt       *time.Time `json:"disposed_at,omitempty"`
	Suggestion       string    `gorm:"-" json:"suggestion,omitempty"` // 处置建议，研判时动态计算，不落库
	CreatedAt        time.Time `gorm:"autoCreateTime" json:"created_at"`
}

func (RiskAssessmentItem) TableName() string {
	return "risk_assessment_items"
}
